lib/srv/forward/sshserver.go

```diff
@@ -851,6 +850,84 @@ func (s *Server) handleConnection(ctx context.Context, chans <-chan ssh.NewChann
 	}
 }
 
+// handleClientChannels handles channel open requests from the remote server.
+func (s *Server) handleClientChannels(ctx context.Context) {
+	forwardedTCPIP := s.remoteClient.HandleChannelOpen(teleport.ChanForwardedTCPIP)
+
+	for {
+		select {
+		case nch := <-forwardedTCPIP:
+			chanCtx, nch := tracessh.ContextFromNewChannel(nch)
+			ctx, span := s.tracerProvider.Tracer("ssh").Start(
+				oteltrace.ContextWithRemoteSpanContext(ctx, oteltrace.SpanContextFromContext(chanCtx)),
+				fmt.Sprintf("ssh.Forward.OpenChannel/%s", nch.ChannelType()),
+				oteltrace.WithSpanKind(oteltrace.SpanKindServer),
+				oteltrace.WithAttributes(
+					semconv.RPCServiceKey.String("ssh.ForwardServer"),
+					semconv.RPCMethodKey.String("OpenChannel"),
+					semconv.RPCSystemKey.String("ssh"),
+				),
+			)
+
+			go func() {
+				defer span.End()
+				s.handleForwardedTCPIPRequest(ctx, nch)
+			}()
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// handleForwardedTCPIPRequest handles remote port forwarding requests.
+func (s *Server) handleForwardedTCPIPRequest(ctx context.Context, nch ssh.NewChannel) {
+	ch, requests, err := nch.Accept()
+	if err != nil {
+		s.log.WithError(err).Errorf("Error accepting %v channel", teleport.ChanForwardedTCPIP)
+		return
+	}
+	go ssh.DiscardRequests(requests)
+	req, err := sshutils.ParseDirectTCPIPReq(nch.ExtraData())
+	if err != nil {
+		s.logErrorToChannel(ch, err, "Unable to parse request.")
+		return
+	}
+
+	// Create context for this channel. This context will be closed when
+	// forwarding is complete.
+	ctx, scx, err := srv.NewServerContext(ctx, s.connectionContext, s, s.identityContext)
+	if err != nil {
+		s.logErrorToChannel(ch, err, "Unable to create connection context.")
+		return
+	}
+	scx.RemoteClient = s.remoteClient
+	scx.ExecType = teleport.ChanDirectTCPIP
+	scx.SrcAddr = sshutils.JoinHostPort(req.Orig, req.OrigPort)
+	scx.DstAddr = sshutils.JoinHostPort(req.Host, req.Port)
+	defer scx.Close()
+
+	ch = scx.TrackActivity(ch)
+
+	// Open a forwarding channel on the client.
+	outCh, outRequests, err := scx.ServerConn.OpenChannel(nch.ChannelType(), nch.ExtraData())
```

Comment:
We should open the channel on the `ServerConn` before accepting the channel from the `remoteClient`; afaik the receiver is supposed to succesfully dial the destination before accepting the channel, or reject the channel with a `ssh.ConnectionFailed` code and an error message; we could forward the error message from the `ServerConn` rejection here, for example.

Reply:
Also, should we filter requests for ports that we know should be open, or are we comfortable leaving that to the client?

Reply:
Even if we wanted to, I don't think there's a way for us to know server-side what ports should be open. We only get the listen addr and port in `tcpip-forward` requests.